airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py

```diff
@@ -0,0 +1,391 @@
+                    "created[lte]": str(int(slice_datetime.timestamp())),
+                    "limit": 100
+                },
+                headers=_AUTHENTICATION_HEADERS,
+            ),
+            _a_response().build()
+        )
+        http_mocker.get(
+            HttpRequest(
+                url="https://api.stripe.com/v1/events",
+                query_params={
+                    "created[gte]": str(int(slice_datetime.timestamp()) + _SECOND_REQUEST),
+                    "created[lte]": str(int(_NOW.timestamp())),
+                    "limit": 100
+                },
+                headers=_AUTHENTICATION_HEADERS,
+            ),
+            _a_response().build()
+        )
+
+        self._read(_config().with_start_date(start_date).with_slice_range_in_days(slice_range.days))
+
+    @HttpMocker()
+    def test_given_http_status_400_when_read_then_stream_is_ignored(self, http_mocker: HttpMocker) -> None:
+        http_mocker.get(
+            HttpRequest(
+                url="https://api.stripe.com/v1/events",
+                query_params=ANY_QUERY_PARAMS,
+                headers=_AUTHENTICATION_HEADERS
+            ),
+            HttpResponse(
+                json.dumps({
+                    "error": {
+                        "code": "parameter_invalid_integer",
+                        "doc_url": "https://stripe.com/docs/error-codes/parameter-invalid-integer",
+                        "message": "Invalid integer: not_an_integer",
+                        "param": "limit",
+                        "request_log_url": "https://dashboard.stripe.com/test/logs/req_yVDTOo9Jrd1suG?t=1701959063",
+                        "type": "invalid_request_error"
+                    }
+                }),
+                400)
+        )
+        output = self._read(_config().with_start_date(_A_START_DATE))
+        assert len(output.get_stream_statuses(_STREAM_NAME)) == 0
+
+    @HttpMocker()
+    def test_given_http_status_401_when_read_then_stream_is_incomplete(self, http_mocker: HttpMocker) -> None:
+        http_mocker.get(
+            HttpRequest(
+                url="https://api.stripe.com/v1/events",
+                query_params=ANY_QUERY_PARAMS,
+                headers=_AUTHENTICATION_HEADERS
+            ),
+            HttpResponse(
+                json.dumps({
+                    "error": {
+                        "message": "Invalid API Key provided: sk_test_*****************************************************mFeM",
+                        "type": "invalid_request_error"
+                    }
+                }),
+                401
+            )
+        )
+        output = self._read(_config().with_start_date(_A_START_DATE), expecting_exception=True)
+        assert output.get_stream_statuses(_STREAM_NAME) == [AirbyteStreamStatus.INCOMPLETE]
+
+    @HttpMocker()
+    def test_given_rate_limited_when_read_then_retry_and_return_records(self, http_mocker: HttpMocker) -> None:
```

Comment:
WDYT about having a wrapper around `unittest.TestCase` (e.g. `BaseFullRefreshTest` and `BaseIncrementalTest`) that provides abstractmethods for test cases that should be covered for every connector (e.g. a rate limiting test case)?

Reply:
This is something I was exploring. The issue I have is that from PyCharm, I can't test a specific method because of that which is annoying for the development loop. Also, these might not be relevant for some APIs because there is no 429/rate limiting which would be odd to force devs to implement it. In the end, I'm not sure it's worth. Do you have a small experiment that would help us confirm or infirm this?

Reply:
I'm not sure I quite understand the pycharm issue. 

I think that for connectors who don't need rate limiting, for example, they could just override the abstractmethod with a dummy method that just does a `pass`. The nice thing about doing that is that it makes it obvious that that particular source does not need to be concerned with rate limiting (it's much harder to look for the absence of a test).

Reply:
Based on the information you added, I misunderstood your point. PyCharm should handle this just fine.

So what I like about this idea is that it's a nice checklist for the dev to make sure he tested all the base use cases. What I don't like about this is the maintenance: if we add a test method in `BaseFullRefreshTest` that is not relevant to my streams and that I have 46 streams like source-stripe has, I would need to update up to 46 tests classes without it providing value to my connector. I can see cases like this pop up somewhat often, one coming soon would be that sources emit state messages on full refresh (but it might not be a feature that makes sense to implement for connector X ASAP). 

So basically, I'm torn about this. I'm wondering if there would be another way to document those features without forcing tests on it. Please, challenge my concerns though because I don't feel 100% convinced about it

Reply:
> I would need to update up to 46 tests classes without it providing value to my connector

this does sound annoying, but I think seeing it as a checklist is right. connector developers should write tests or acknowledge they should've written one but explicitly skipped it for a given stream.

^feel free to challenge that statement, but I like the direction of enforced checklists. it seems more scalable than relying on a reviewer to verify a checklist has been followed

Reply:
An issue has been created for this [here](https://github.com/airbytehq/airbyte/issues/33417)

Reply:
I suppose developers could also make a `BaseStripeFullRefreshTest(BaseFullRefreshTest)` that indicates which ones aren't needed for Stripe, and they inherit from that for the 46 streams.